trieste/acquisition/rule.py

```diff
@@ -1097,6 +1104,225 @@ def state_func(
+
+    @dataclass(frozen=True)
+    class State:
+        """The acquisition state for the :class:`TURBO` acquisition rule."""
+
+        acquisition_space: Box
+        """ The search space. """
+
+        L: float
+        """ Length of the trust region (before standardizing by model lengthscales) """
+
+        failure_counter: int
+        """ Number of consecutive failures (reset if we see a success). """
+
+        success_counter: int
+        """ Number of consecutive successes (reset if we see a failure).  """
+
+        y_min: TensorType
+        """ The minimum observed value. """
+
+        def __deepcopy__(self, memo: dict[int, object]) -> TURBO.State:
+            box_copy = copy.deepcopy(self.acquisition_space, memo)
+            return TURBO.State(
+                box_copy, self.L, self.failure_counter, self.success_counter, self.y_min
+            )
+
+    def __init__(
+        self,
+        search_space: SearchSpace,
+        num_query_points: int = 1,
+        L_min: Optional[float] = None,
+        L_init: Optional[float] = None,
+        L_max: Optional[float] = None,
+        success_tolerance: int = 3,
+        failure_tolerance: Optional[int] = None,
+        num_samples: Optional[int] = None,
+    ):
+        """
+        Note that the optional parameters are set by a heuristic if not given by the user.
+
+        :param search_space: The search space.
+        :param num_query_points: The number of points in a batch.
+        :param L_min: Minimum allowed length of the trust region.
+        :param L_init: Initial length of the trust region.
+        :param L_max: Maximum allowed length of the trust region.
+        :param success_tolerance: Number of consecutive successes before changing region size.
+        :param failure tolerance: Number of consecutive failures before changing region size.
+        :param num_samples: Size of sample used for Thompson sampling.
+        """
+
+        if not num_query_points > 0:
+            raise ValueError(f"Num query points must be greater than 0, got {num_query_points}")
+
+        if num_query_points > 1:
+            raise NotImplementedError(
+                f"TURBO does not yet support batches greater than 1, got {num_query_points}"
+            )
+
+        # implement heuristic defaults for trust region params if not specified by user
+        if failure_tolerance is None:
+            failure_tolerance = math.ceil(search_space.dimension / num_query_points)
+        search_space_max_width = tf.reduce_max(search_space.upper - search_space.lower)
+        if L_min is None:
+            L_min = (0.5**7) * search_space_max_width
+        if L_init is None:
+            L_init = 0.8 * search_space_max_width
+        if L_max is None:
+            L_max = 1.6 * search_space_max_width
+        if num_samples is None:
+            num_samples = tf.clip_by_value(100 * search_space.dimension, 5_000, 1_000_000)
```

Comment:
in the paper I think num_sample is min(100d, 5000) since it's exact TS

Reply:
yep fixed